OpenTrons/opentrons-api
# fix(robot-server): Fix protocols being stored in the wrong directory

# Overview

Closes RSS-469 and associated issues, like RQA-2294 and RQA-2295.

# Test Plan

I've confirmed that on `edge`, these steps reproduce the error in described in RQA-2295, and with this branch, they do not:

1. Put a lot of different protocols ([≥20 if you're using a Flex](https://github.com/Opentrons/oe-core/blob/9942b8aa873109af2000e42929dd057ab6f4d28a/layers/meta-opentrons/recipes-robot/opentrons-robot-server/files/opentrons-robot-server.service#L16) or [≥5 if you're using a dev server](https://github.com/Opentrons/opentrons/blob/8f4b709fd3cd865ea427e6e5dd0861f7397f75a4/robot-server/robot_server/settings.py#L88)) on a v7.1.0 server.
2. Upgrade to this branch.
3. Upload a new protocol. This will auto-delete one of the old ones to make room.
4. Downgrade back to v7.1.0.
5. Make sure all the original protocols are still available (either by manually issuing HTTP GET requests or by clicking through the Opentrons App).

# Changelog

robot-server stores stuff in a `protocols/` directory and in a `robot_server.db` file. They need to be kept in sync.

The bug, introduced in #14383, is that after a migration to schema 3, we were still using the `protocols/` directory from schema 2. So the server was working from the new `robot_server.db` file but the old `protocols/` directory. When you downgraded back to v7.1.0, the server would hit errors from the old `protocols/` directory now being out of sync with the old `robot_server.db` file. This was caused by a pretty embarrassing mistake, and the fix is one line.

By fixing that, we expose another problem. The `POST /settings/reset` endpoint lets you clear deck configuration. But it can only do that after the persistence layer has initialized. And that endpoint is supposed to be highly available; it's supposed to work even if the database is corrupt, or taking a million hours to migrate. This PR's quick and dirty solution is to return HTTP 503 if you asked to clear deck configuration but things aren't yet initialized far enough to do that.

# Review requests

1. Does my treatment of `POST /settings/reset` seem sensible? It's not really clear to me how that endpoint should ideally behave. Should it block until it's able to reset everything that was requested? Should `deckConfiguration` even be an option there, as opposed to it being a separate endpoint like `DELETE /deck_configuration`?
2. What do we want to do with the internal robots that have already been infected with this borked migration? Unfortunately, I don't see a sensible way of recovering from it. We can reset those robots' runs and protocols, but then we lose some QA coverage of the migration. I could run around the office and manually fix things up, I guess? There are "only" 10-15 robots that have gone through this update so far, right?

# Risk assessment

Medium. As the bugs show, this upgrade/downgrade stuff is not easy to catch with our existing automated tests.
